BUG: complex repr has extra spaces, missing +

ahaldane opened the pull request:
Fixes #10355
Fixes #10356 

This still needs a release note, and some discussion of the right behavior.

In particular, for 1.13 and in legacy mode we have
```python
>>> np.array([1.0 + 1.0j, 1.123456789 + 1.123456789j], dtype='c16')
array([ 1.00000000+1.j        ,  1.12345679+1.12345679j])
```

But after this PR 1.14 will have
```python
>>> np.array([1.0 + 1.0j, 1.123456789 + 1.123456789j], dtype='c16')
array([1.00000000+1.00000000j, 1.12345679+1.12345679j])
```

In other words the imaginary part now always prints an equal number of digits for all elements, just like the real part did before. Does that sound reasonable?

ahaldane committed:
BUG: repr of negative zeros sometimes has an extra space

ahaldane commented:
By the way the changes allowing precision to be `None` are not necessary, just a little tidying up.

eric-wieser commented:
What does `floatmode='unique'` do for this case?

ahaldane commented:
@eric-wieser 
Normally for floats, `floatmode='unique'` prints only the number of digits needed to specify the value. Specifying `floatmode='unique'` is actually exactly equivalent to specifying `floatmode='maxprec', precision=None` for plain floats. And to further explain *that*, `floatmode='maxprec'` means "print at most `precision` digits (or infinite digits if precision is None), but stopping if the vaue has been uniquely specified.

But what I've done in this PR is special-case complex arrays, so that `floatmode='unique'` is equivalent to ``floatmode='maxprec_equal', precision=None` for the real and imag parts., which makes all the values print the same number of digits.

pv commented:
Does this also stop printing numbers as `1.00000000+1.0           j` where there were spaces before the `j`, or was that a separate bug?

ahaldane commented:
@pv yes, those spaces are removed in both legacy mode and non-legacy mode.

eric-wieser commented:
To elaborate on the above comment I forgot to click submit on - I think providing some way to only show the relevant digits of a `complex` number is just as useful as having it for a float:

```
>>> np.array([0.5, 0.0005])
array([0.5   , 0.0005])
>>> np.array([0.5, 0.0005]) * 1+1j
array([0.5   +0.5   j, 0.0005+0.0005j])  # currently
array([0.5000+0.5000j, 0.0005+0.0005j])  # this patch
array([0.5   +0.5j   , 0.0005+0.0005j])  # my proposal for unique mode

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1075,15 +1090,44 @@ def __init__(self, x, precision, floatmode, suppress_small,
         i = self.imag_format(x.imag)
-        return r + i + 'j'
+
+        if self.legacy is not False:
```
Should always compare against `"1.13"`

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1035,7 +1047,10 @@ def format_float_positional(x, precision=None, unique=True,
+    if precision is None:
+        precision = -1
+    elif precision < 0:
+        raise ValueError("precision must be >= 0")
```
Why are we now forbidding `-1` passed directly? If we do it here, shouldn't you do it below too?

ahaldane replied:
I consider the `-1` to be an ugly internal detail of the Python-C interface for float printing, forced by the behavior of optional args in `PyArgParseTuple`, which I want to hide. It's much clearer if the user should supply `None` if they mean unlimited precision. (Other `multiarraymodule.c` functions also have this `-1` quirk, if I remember correctly).

Where do you mean by "below"? I thought I caught all the places the user can supply precision.

eric-wieser replied:
The padding arguments on the next line

eric-wieser replied:
`pad_left = -1 if pad_left is None else pad_left`, for example.

Perhaps a simple converter function in C would be worthwhile here?
```
int positive_or_none(PyObject *obj, void* ctx) {
    int *result = ctx;
    if (obj == Py_None) {
        *result = -1;
        return 0;
    }
    // convert to int, check sign
}
```

ahaldane commented:
Updated, preserving `unique` mode as you suggested. The behavior now is:
```python
[1]: a = np.random.rand(10) + 1+1j
[2]: a[0] = 1.0 + 1.1234j
[3]: a  # "maxprec" repr mode by default
array([1.        +1.1234j, 1.52782481+1.j    , 1.30953621+1.j    ,
       1.78469893+1.j    , 1.73193001+1.j    , 1.50286996+1.j    ,
       1.71633843+1.j    , 1.39270544+1.j    , 1.67535861+1.j    ,
       1.08396307+1.j    ])
[4]: np.set_printoptions(floatmode='maxprec_equal')
[5]: a
array([1.00000000+1.1234j, 1.52782481+1.0000j, 1.30953621+1.0000j,
       1.78469893+1.0000j, 1.73193001+1.0000j, 1.50286996+1.0000j,
       1.71633843+1.0000j, 1.39270544+1.0000j, 1.67535861+1.0000j,
       1.08396307+1.0000j])
```

Seems good to me.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1075,15 +1090,33 @@ def __init__(self, x, precision, floatmode, suppress_small,
         i = self.imag_format(x.imag)
-        return r + i + 'j'
+
+        # add the 'j' at the first space in i
```
*before, not at

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1075,15 +1090,33 @@ def __init__(self, x, precision, floatmode, suppress_small,
+        # since we should not have any spaces in the complex values, we can't
+        # use "maxprec" or "unique" float-modes since they will shorten the
+        # fractional parts before the 'j'. In both cases use the
+        # 'maxprec_equal' mode instead.
```
Comment seems out of date

ahaldane commented:
Fixed up. 

I also realized my code for determining the `j` index was buggy if there was left-padding, fixed now.

eric-wieser commented on doc/release/1.14.1-notes.rst:
```diff
@@ -0,0 +1,22 @@
+==========================
+NumPy 1.14.1 Release Notes
+==========================
```
@charris: Can you make a separate PR based upon 5f01e54b20e38d483e8bab31bf5f953a860fe8d3 that adds these release notes?

eric-wieser replied:
Actually - @ahaldane, if you rebase on that commit and split adding/modifying this file in two, that would have the same effect.

ahaldane replied:
Did you link the wrong commit? That commit doesn't touch the template part of the release note. Was that the intention?

eric-wieser replied:
That commit is the original common ancestor between 1.14 and master that I've been basing all my backportable PRs on

eric-wieser replied:
At risk of being nitpicky, I'd squash your 1st and 3rd commits together, and put them after the template one - that way we can base backportable PRs on top of the template commit

charris replied:
`cherry-pick` is so easy, I really don't see the point of using a common base, and If you want to use the same branch, just rebase (onto), force push, and do another PR. If you look at the graph, basing on a point in the past also just fattens up the number of lines from the past.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -759,17 +768,21 @@ def __init__(self, data, precision, floatmode, suppress_small, sign=False,
         if self._legacy == '1.13':
-            sign = '-' if data.shape == () else ' '
+            # legacysign is a secret FloatingFormat option that adds a sign
+            # even in legacy mode, needed for imag of complex.
```
How was this handled before?

ahaldane replied:
Right, in 1.13 the `sign` option took care of it. But now in 1.14 we have taken over and redefined `sign`, and since it is now ignored in legacy mode we need another way of forcing the '+'.

eric-wieser replied:
I think this would also work:
```
if self._legacy == '1.13':
    # when not 0d, legacy does not support '-'
    if data.shape != () and sign == '-':
        sign = ' '
```

ahaldane replied:
True. I considered something similar but I was worried about people using `sign='+', legacy='1.13'` at the same time, and getting confused at non-legacy-like output since the `+` option didn't exist in 1.13.

But on second thought I'll go with your suggestion.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1075,15 +1087,26 @@ def __init__(self, x, precision, floatmode, suppress_small,
+
+        # add the 'j' before the terminal whitespace in i
+        sp = len(i.rstrip())
+        i = i[:sp] + 'j' + i[sp:]
```
Could also use `(i.rstrip() + 'j').ljust(len(i) + 1)`

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -427,28 +427,37 @@ def test_bool_spacing(self):
         assert_equal(repr(np.array(1.)), 'array(1.0)')
         assert_equal(repr(np.array([0.])), 'array([ 0.])')
+        assert_equal(repr(c),
+            "array([ 1.00000000+1.j        ,  1.12345679+1.12345679j])")
```
The legacy output was just bad!

ahaldane committed:
MAINT: Add 1.14.1 release note template

ahaldane committed:
BUG: complex repr has extra spaces, missing +

Fixes #10355
Fixes #10356

ahaldane committed:
BUG: correctly add initial space in legacy printing

Fixes #10383

ahaldane commented:
Added a commit to fix #10383

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -861,7 +863,8 @@ def fillFormat(self, data):
         if all(finite_vals >= 0) and self.sign == ' ':
-            self.pad_left += 1
+            if self._legacy != '1.13':
+                self.pad_left += 1
```
Would read better with the `if`s inverted (and comment moved within the inner one), I think

ahaldane commented:
Added a commit for another little bug I found: An extra space could be added for a negative 0 ('-0.`):
```
>>> np.set_printoptions(sign=' ')
>>> x
array([  0.,  -0.])
```
fixed by changing the `x >= 0` test to `not np.signbit(x)`

eric-wieser commented on doc/release/1.14.1-notes.rst:
```diff
@@ -6,6 +6,8 @@ This is a bugfix release for some problems found since 1.14.0. The most
 
 
+important fixes are 
+important fixes are to the spacing in the str and repr of complex values.
```
Rebase mistake here

charris commented:
@ahaldane , @eric-wieser  Ready?

ahaldane commented:
I think so, I have nothing more to add.

ahaldane commented on numpy/core/arrayprint.py:
```diff
@@ -868,7 +878,11 @@ def fillFormat(self, data):
             int_part, frac_part = zip(*(s.split('.') for s in strs))
-            self.pad_left = max(len(s) for s in int_part)
+            if self._legacy == '1.13':
+                self.pad_left = 1 + max(len(s.lstrip('-').lstrip('+'))
```
Oh, so it is! I thought it was a string, not a set of chars.

eric-wieser replied:
Yeah, that would have been a more useful API, but oh well

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -880,9 +893,10 @@ def fillFormat(self, data):
+        if self._legacy != '1.13':
+            # account for sign = ' ' by adding one to pad_left
+            if not any(np.signbit(finite_vals)) and self.sign == ' ':
+                    self.pad_left += 1
```
nit: weird indentation

ahaldane replied:
fixed

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -880,9 +893,10 @@ def fillFormat(self, data):
-            self.pad_left += 1
+        if self._legacy != '1.13':
+            # account for sign = ' ' by adding one to pad_left
+            if not any(np.signbit(finite_vals)) and self.sign == ' ':
```
Is there a test for signed zero? Nice catch!

ahaldane replied:
Yup, I added the following test in lines 444 and 452 of test_arrayprint:
```python
assert_equal(repr(np.array([0., -0.])), 'array([ 0., -0.])')
```

charris commented:
So in it goes. Thanks Allan, Eric.
